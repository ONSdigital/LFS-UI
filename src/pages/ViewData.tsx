import React, {Component} from "react";
import DocumentTitle from "react-document-title";
import {VariableDefinitionTable} from "../components/ViewDataTables/VariableDefinitionTable";
import {ValueLabelsTable} from "../components/ViewDataTables/ValueLabelsTable";
import {ONSTabs} from "../components/ONS_DesignSystem/ONSTabs";

interface State {
    VarDefData: [] | null,
    table: any | null
    tabItems: any[]
}

export class ViewData extends Component <{}, State> {
    displayName = ViewData.name;

    constructor(props: any) {
        super(props);
        this.state = {
            VarDefData: [],
            table: null,
            tabItems: this.tabsSelection
        };
    }

    componentDidMount(): void {
        this.handleTableChange("Variable Definitions");
    }

    handleTableChange = (tableName: string) => {
        let table = null;

        switch (tableName) {
            case "Value Labels":
                table = <ValueLabelsTable/>;
                break;
            case "Variable Definitions":
                table = <VariableDefinitionTable/>;
                break;
        }
        this.setState({table: table});

        // Set Previous active tab as not active
        let items = this.state.tabItems;
        let item = items.find(x => x.active);
        if (item !== undefined) {
            item.active = false;
        }

        // Find new selected tab and set as active
        item = items.find(x => x.name === tableName);
        if (item !== undefined) {
            item.active = true;
        }
    };

    tabsSelection = [
        {name: "Variable Definitions", active: true},
        {name: "Value Labels", active: false}
    ];

    render() {
        return (
            <DocumentTitle title={"LFS View Data"}>
                <div className={"container"}>
                    <br/>
                    <ONSTabs label={"Select Table"} items={this.state.tabItems} onClick={this.handleTableChange}/>
                    {this.state.table}
                </div>
            </DocumentTitle>
        );
    }
}
