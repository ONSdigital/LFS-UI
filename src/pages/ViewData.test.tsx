import React from "react";
import {cleanup, fireEvent, render} from "@testing-library/react";

import {BrowserRouter} from "react-router-dom";

import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import fetch from "../tests/setup/__mocks__/fetch.js";
import {ViewData} from "./ViewData";
import flushPromises from "../tests/util/flushPromises";


// @ts-ignore
global.fetch = fetch;


describe("View Data Page", () => {
    Enzyme.configure({adapter: new Adapter()});

    beforeEach(() => {
        process.env.NODE_ENV = "test";
    });

    afterEach(() => {
        cleanup();
    });


    function wrapper(render: any) {
        return render(
            <BrowserRouter>
                <ViewData/>
            </BrowserRouter>
        );
    }

    it("renders the page", async () => {
        expect(wrapper(render)).toMatchSnapshot();
    });

    it("should display the Variable Definitions by default", async () => {
        const {queryAllByTestId, getByText} = wrapper(render);

        await flushPromises();

        expect(getByText(/DVHRPNUM/i));

        let list = queryAllByTestId(/table-row/i);

        expect(list.length).toEqual(2);
    });

    test("When you click on the tabs it should change the Table to the selected item.\n", async () => {
        const {queryAllByTestId, getByText} = wrapper(render);

        await flushPromises();

        await fireEvent.click(getByText(/Value Labels/i));

        await flushPromises();

        let list = queryAllByTestId(/table-row/i);

        expect(list.length).toEqual(4);
    });

    /*
    Variable Definitions Table tests
     */

    it("should loads all Variable Definitions in the list", async () => {
        const {queryAllByTestId} = wrapper(render);

        await flushPromises();

        let list = queryAllByTestId(/table-row/i);

        expect(list.length).toEqual(2);
    });

    it("should filter list by the Variable Name entered in the Search field for Variable Definitions", async () => {
        const {queryAllByTestId, getByLabelText, getByText, getAllByText} = wrapper(render);

        await flushPromises();

        await fireEvent.change(getByLabelText(/Filter by Variable Name/i), {
            target: {value: "IOUT"}
        });

        await flushPromises();

        let list = queryAllByTestId(/table-row/i);

        expect(list.length).toEqual(1);

        expect(getAllByText(/IOUTCOME/i)).toBeDefined();

        // Clear List back to default
        await fireEvent.click(getByText(/Clear/i));
        await flushPromises();
        list = queryAllByTestId(/table-row/i);
        expect(list.length).toEqual(2)
    });


    test("When you click on a Row the previous definitions for a variable should be shown", async () => {
        const {getByText, getAllByText} = wrapper(render);

        await flushPromises();

        await fireEvent.click(getByText(/IOUTCOME/i));

        await flushPromises();

        expect(getAllByText(/Previous Definition description/i));
    });

    test("When you click on a Row and there is No Previous Metadata a message is shown", async () => {
        const {getByText, getAllByText} = wrapper(render);

        await flushPromises();

        await fireEvent.click(getByText(/DVHRPNUM/i));

        await flushPromises();

        expect(getAllByText(/No Previous Metadata/i));
    });

    it("should display a message when no  are found", async () => {
        const {queryAllByTestId} = wrapper(render);

        await flushPromises();

        let list = queryAllByTestId(/table-row/i);

        expect(list.length).toEqual(2);
    });

    it("should display a message when no Variable Definitions are found", async () => {
        process.env.NODE_ENV = "noData";

        const {getByText} = wrapper(render);

        await flushPromises();

        expect(getByText(/No Variable Definitions found/i)).toBeDefined();
    });

    it("should display an error when the request fails to get the Variable Definitions", async () => {
        process.env.NODE_ENV = "returnError";

        const {getByText} = wrapper(render);

        await flushPromises();

        expect(getByText(/Error occurred while getting Variable Definitions/i)).toBeDefined();
    });

    /*
    Value Label Table tests
     */

    it("should loads all Value Labels in the list", async () => {
        const {getByText, queryAllByTestId} = wrapper(render);

        await flushPromises();

        await fireEvent.click(getByText(/Value Labels/i));

        await flushPromises();

        let list = queryAllByTestId(/table-row/i);

        expect(list.length).toEqual(4);
    });

    it("should filter list by the Variable Name entered in the Search field for Value Labels", async () => {
        const {queryAllByTestId, getByLabelText, getByText, getAllByText} = wrapper(render);

        await flushPromises();

        await fireEvent.click(getByText(/Value Labels/i));

        await flushPromises();

        await fireEvent.change(getByLabelText(/Filter by Variable Name/i), {
            target: {value: "ILLS"}
        });

        await flushPromises();

        let list = queryAllByTestId(/table-row/i);

        expect(list.length).toEqual(3);

        expect(getAllByText(/ILLST17/i)).toBeDefined();

        // Clear List back to default
        await fireEvent.click(getByText(/Clear/i));
        await flushPromises();
        list = queryAllByTestId(/table-row/i);
        expect(list.length).toEqual(4)
    });

    it("should display a message when no Value Labels are found", async () => {
        process.env.NODE_ENV = "noData";

        const {getByText} = wrapper(render);

        await flushPromises();

        await fireEvent.click(getByText(/Value Labels/i));

        await flushPromises();

        expect(getByText(/No Value Labels found/i)).toBeDefined();
    });

    it("should display an error when the request fails to get the Value Labels", async () => {
        process.env.NODE_ENV = "returnError";

        const {getByText} = wrapper(render);

        await flushPromises();

        await fireEvent.click(getByText(/Value Labels/i));

        await flushPromises();

        expect(getByText(/Error occurred while getting Value Labels/i)).toBeDefined();
    });

});