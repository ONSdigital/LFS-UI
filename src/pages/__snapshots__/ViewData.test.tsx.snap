// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View Data Page renders the page 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <br />
        <div
          class="tabs"
        >
          <label
            class="label"
          >
            Select Table
          </label>
          <ul
            class="tabs__list tabs__list--row"
            role="tablist"
          >
            <li
              class="tab__list-item tab__list-item--row"
              id="tabIdItem0"
            >
              <a
                aria-selected="true"
                class="tab tab--row tab--selected"
                href="#tabId"
                role="tab"
                tabindex="1"
              >
                Variable Definitions
              </a>
            </li>
            <li
              class="tab__list-item tab__list-item--row"
              id="tabIdItem1"
            >
              <a
                aria-selected="false"
                class="tab tab--row "
                href="#tabId"
                role="tab"
                tabindex="1"
              >
                Value Labels
              </a>
            </li>
          </ul>
        </div>
        <div
          class="grid--flex "
        >
          <p
            class="field"
          >
            <label
              class="label"
              for="variableDefinitionSearch"
            >
              Filter by Variable Name
            </label>
            <input
              class="input input--text input-type__input u-mr-xs u-mb-xs input--w-10"
              id="variableDefinitionSearch"
              type="text"
              value=""
            />
          </p>
          <button
            class="btn btn--small u-flex-align-bottom u-mb-xs btn--secondary "
            type="submit"
          >
            <span
              class="btn__inner"
            >
              Clear
            </span>
          </button>
        </div>
        <table
          class="table table--sortable"
          data-aria-asc="ascending"
          data-aria-desc="descending"
          data-aria-sort="Sort by"
          id="basic-table"
        >
          <thead
            class="table__head"
          >
            <tr
              class="table__row"
            >
              <th
                aria-sort="none"
                class="table__header "
                scope="col"
              />
              <th
                aria-sort="none"
                class="table__header "
                scope="col"
              >
                <button
                  aria-label="Sort by Variable"
                  class="table__sort-button"
                  data-index="0"
                  type="button"
                >
                  Variable
                </button>
              </th>
              <th
                class="table__header "
                scope="col"
              >
                Description
              </th>
              <th
                aria-sort="none"
                class="table__header "
                scope="col"
              >
                <button
                  aria-label="Sort by Valid From"
                  class="table__sort-button"
                  data-index="2"
                  type="button"
                >
                  Valid From
                </button>
              </th>
              <th
                class="table__header "
                scope="col"
              >
                Editable
              </th>
              <th
                class="table__header "
                scope="col"
              >
                Imputation
              </th>
              <th
                class="table__header "
                scope="col"
              >
                DV
              </th>
              <th
                class="table__header "
                scope="col"
              >
                Type
              </th>
              <th
                class="table__header "
                scope="col"
              >
                Length
              </th>
            </tr>
          </thead>
          <tbody
            class="table__body expandedRowEnabled "
          >
            <tr>
              <td
                class="table__cell "
                colspan="9"
              >
                <div
                  class="panel panel--info panel--simple "
                >
                  <div
                    class="panel__body"
                  >
                    <p>
                      No Variable Definitions matching this criteria
                    </p>
                  </div>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
        <div
          hidden=""
        >
          <nav
            aria-label="Pagination (Page 1 of 0)"
            class="pagination "
          >
            <div
              class="pagination__position"
            >
              Page 
              1
               of 
              0
            </div>
            <ul
              class="pagination__items"
              style="margin-left: 0px;"
            >
              <li
                class="pagination__item pagination__item--current"
                style="cursor: pointer;"
              >
                <button
                  aria-label="Go to page 1"
                  class="pagination__link"
                  id="link"
                  style="cursor: pointer; outline: 0; text-decoration: underline;"
                >
                  1
                </button>
              </li>
              <li
                class="pagination__item pageination__item--gap"
              >
                …
              </li>
              <li
                class="pagination__item pageination__item--gap"
              >
                …
              </li>
              <li
                class="pagination__item "
                style="cursor: pointer;"
              >
                <button
                  aria-label="Go to page 0"
                  class="pagination__link"
                  id="link"
                  style="cursor: pointer; outline: 0; text-decoration: underline;"
                >
                  0
                </button>
              </li>
            </ul>
          </nav>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <br />
      <div
        class="tabs"
      >
        <label
          class="label"
        >
          Select Table
        </label>
        <ul
          class="tabs__list tabs__list--row"
          role="tablist"
        >
          <li
            class="tab__list-item tab__list-item--row"
            id="tabIdItem0"
          >
            <a
              aria-selected="true"
              class="tab tab--row tab--selected"
              href="#tabId"
              role="tab"
              tabindex="1"
            >
              Variable Definitions
            </a>
          </li>
          <li
            class="tab__list-item tab__list-item--row"
            id="tabIdItem1"
          >
            <a
              aria-selected="false"
              class="tab tab--row "
              href="#tabId"
              role="tab"
              tabindex="1"
            >
              Value Labels
            </a>
          </li>
        </ul>
      </div>
      <div
        class="grid--flex "
      >
        <p
          class="field"
        >
          <label
            class="label"
            for="variableDefinitionSearch"
          >
            Filter by Variable Name
          </label>
          <input
            class="input input--text input-type__input u-mr-xs u-mb-xs input--w-10"
            id="variableDefinitionSearch"
            type="text"
            value=""
          />
        </p>
        <button
          class="btn btn--small u-flex-align-bottom u-mb-xs btn--secondary "
          type="submit"
        >
          <span
            class="btn__inner"
          >
            Clear
          </span>
        </button>
      </div>
      <table
        class="table table--sortable"
        data-aria-asc="ascending"
        data-aria-desc="descending"
        data-aria-sort="Sort by"
        id="basic-table"
      >
        <thead
          class="table__head"
        >
          <tr
            class="table__row"
          >
            <th
              aria-sort="none"
              class="table__header "
              scope="col"
            />
            <th
              aria-sort="none"
              class="table__header "
              scope="col"
            >
              <button
                aria-label="Sort by Variable"
                class="table__sort-button"
                data-index="0"
                type="button"
              >
                Variable
              </button>
            </th>
            <th
              class="table__header "
              scope="col"
            >
              Description
            </th>
            <th
              aria-sort="none"
              class="table__header "
              scope="col"
            >
              <button
                aria-label="Sort by Valid From"
                class="table__sort-button"
                data-index="2"
                type="button"
              >
                Valid From
              </button>
            </th>
            <th
              class="table__header "
              scope="col"
            >
              Editable
            </th>
            <th
              class="table__header "
              scope="col"
            >
              Imputation
            </th>
            <th
              class="table__header "
              scope="col"
            >
              DV
            </th>
            <th
              class="table__header "
              scope="col"
            >
              Type
            </th>
            <th
              class="table__header "
              scope="col"
            >
              Length
            </th>
          </tr>
        </thead>
        <tbody
          class="table__body expandedRowEnabled "
        >
          <tr>
            <td
              class="table__cell "
              colspan="9"
            >
              <div
                class="panel panel--info panel--simple "
              >
                <div
                  class="panel__body"
                >
                  <p>
                    No Variable Definitions matching this criteria
                  </p>
                </div>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
      <div
        hidden=""
      >
        <nav
          aria-label="Pagination (Page 1 of 0)"
          class="pagination "
        >
          <div
            class="pagination__position"
          >
            Page 
            1
             of 
            0
          </div>
          <ul
            class="pagination__items"
            style="margin-left: 0px;"
          >
            <li
              class="pagination__item pagination__item--current"
              style="cursor: pointer;"
            >
              <button
                aria-label="Go to page 1"
                class="pagination__link"
                id="link"
                style="cursor: pointer; outline: 0; text-decoration: underline;"
              >
                1
              </button>
            </li>
            <li
              class="pagination__item pageination__item--gap"
            >
              …
            </li>
            <li
              class="pagination__item pageination__item--gap"
            >
              …
            </li>
            <li
              class="pagination__item "
              style="cursor: pointer;"
            >
              <button
                aria-label="Go to page 0"
                class="pagination__link"
                id="link"
                style="cursor: pointer; outline: 0; text-decoration: underline;"
              >
                0
              </button>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
